internals/overlord/checkstate/handlers.go

```diff
@@ -39,63 +107,46 @@ func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tombpkg.Tomb) erro
 	}
 
 	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+
+	// Retrieve checkData and its channels
+	m.checksLock.Lock()
+	m.ensureCheck(config.Name)
+	checkData := m.checks[details.Name]
+	refresh := checkData.refresh
+	result := checkData.result
+	m.checksLock.Unlock()
+
 	ticker := time.NewTicker(config.Period.Value)
 	defer ticker.Stop()
 
 	chk := newChecker(config)
+
+	ctx := &checkContext{
+		task:       task,
+		tomb:       tomb,
+		chk:        chk,
+		changeID:   changeID,
+		config:     config,
+		details:    &details,
+		sendResult: false,
+		result:     result,
+	}
+
 	for {
 		select {
-		case <-ticker.C:
-			err := runCheck(tomb.Context(nil), chk, config.Timeout.Value)
-			if !tomb.Alive() {
-				return checkStopped(config.Name, task.Kind(), tomb.Err())
+		case <-refresh:
+			// Reset ticker on refresh.
+			ticker.Reset(config.Period.Value)
+			ctx.sendResult = true
+			err := m.performCheckAndSendResult(ctx)
+			if err != nil {
+				return err
 			}
+		case <-ticker.C:
+			err := m.performCheckAndSendResult(ctx)
```

Comment:
Here's what I was thinking with checking the refresh channel:

```suggestion
			err := m.performCheckAndSendResult(ctx)
			select {
			case <-refresh: // if refresh requested while running check, send result
			  result <- err
			default: // otherwise don't send result
			}
```